discovery/src/main/java/com/alibaba/fescar/discovery/registry/consul/ConsulRegistryServiceImpl.java

```diff
@@ -0,0 +1,336 @@
+
+    private static ConcurrentMap<String, List<InetSocketAddress>> clusterAddressMap = null;
+    private static ConcurrentMap<String, Set<ConsulListener>> listenerMap = null;
+    private static ExecutorService notifierExecutor = null;
+    private static ConcurrentMap<String, ConsulNotifier> notifiers = null;
+
+    private static final int THREAD_POOL_NUM = 1;
+
+    /**
+     * default tcp check interval
+     */
+    private static final String DEFAULT_CHECK_INTERVAL = "10s";
+    /**
+     * default tcp check timeout
+     */
+    private static final String DEFAULT_CHECK_TIMEOUT = "1s";
+    /**
+     * default deregister critical server after
+     */
+    private static final String DEFAULT_DEREGISTER_TIME = "20s";
+    /**
+     * default watch timeout in second
+     */
+    private static final int DEFAULT_WATCH_TIMEOUT = 60;
+
+
+    private ConsulRegistryServiceImpl() {
+    }
+
+    /**
+     * get instance of ConsulRegistryServiceImpl
+     *
+     * @return instance
+     */
+    public static ConsulRegistryServiceImpl getInstance() {
+        if (null == instance) {
+            synchronized (ConsulRegistryServiceImpl.class) {
+                if (null == instance) {
+                    //initial the capacity with 8
+                    clusterAddressMap = new ConcurrentHashMap<>(1 << 3);
+                    listenerMap = new ConcurrentHashMap<>(1 << 3);
+                    notifiers = new ConcurrentHashMap<>(1 << 3);
+                    notifierExecutor = new ThreadPoolExecutor(THREAD_POOL_NUM, THREAD_POOL_NUM,
+                        Integer.MAX_VALUE, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>(),
+                        new NamedThreadFactory("fescar-consul-notifier", THREAD_POOL_NUM));
+                    instance = new ConsulRegistryServiceImpl();
+
+                }
+            }
+        }
+        return instance;
+    }
+
+    @Override
+    public void register(InetSocketAddress address) throws Exception {
+        NetUtil.validAddress(address);
+        getConsulClient().agentServiceRegister(createService(address));
+    }
+
+    @Override
+    public void unregister(InetSocketAddress address) throws Exception {
+        NetUtil.validAddress(address);
+        getConsulClient().agentServiceDeregister(createServiceId(address));
+    }
+
+    @Override
+    public void subscribe(String cluster, ConsulListener listener) throws Exception {
+        //1.add listener to subscribe list
+        listenerMap.putIfAbsent(cluster, new HashSet<>());
+        listenerMap.get(cluster).add(listener);
+        //2.get healthy services
+        Response<List<HealthService>> response = getHealthyServices(cluster, -1, DEFAULT_WATCH_TIMEOUT);
+        //3.get current consul index.
+        Long index = response.getConsulIndex();
+        ConsulNotifier notifier = notifiers.computeIfAbsent(cluster, k -> new ConsulNotifier(cluster, index));
+        //4.run notifier
+        notifierExecutor.submit(notifier);
```

Comment:
when it fail how to do ？I think we can use the future pattern so that all implementation classes can be applied。we can also extracting common constants。

Reply:
i will improve this in next PR.

Reply:
expect